docs/adr/2023-09-08_026-network-resilience.md

```diff
@@ -0,0 +1,68 @@
+  * Our [FireForget](https://github.com/input-output-hk/hydra/blob/8a8e0829964132bde8949e5249a1ab303af92fb8/hydra-node/src/Hydra/Network/Ouroboros/Type.hs#L31) ignores connections/disconnections
+  * Ouroboros/typed-protocols provides enough machinery to implement a reliable broadcast protocol, for example by reusing existing `KeepAlive` protocol and building a more robust point-to-point protocol than what we have now
+  * There is a minor limitation, namely that the subscription mechanism does not handle connections invidually, but as a set of equivalent point-to-point full duplex connections whose size (valency) needs to be maintained at a certain threshold, which means that unless backed in the protocol itself, protocol state-machine and applications are not aware of the identity of the remote peer
+* We have built our `Network` infrastructure over the concept of relatively independent layers, each implementing a similar interface with different kind of messages, to `broadcast` messages to all peers and be notified of incoming messages through a `callback`.
+  * This pipes-like abstraction allows us to compose our network stack like:
+
+    ```
+     withAuthentication (contramap Authentication tracer) signingKey otherParties $
+      withHeartbeat nodeId connectionMessages $
+        withOuroborosNetwork (contramap Network tracer) localhost peers
+    ```
+
+  * This has the nice property that we can basically swap the lower layers should we need to, for example to use [UDP](https://github.com/input-output-hk/hydra/blob/abailly-iohk/multi-node-udp/hydra-node/src/Hydra/Network/UDP.hs), or add other layers for example to identify [Multiple Heads](https://github.com/input-output-hk/hydra/blob/abailly-iohk/multi-node-udp/hydra-node/src/Hydra/Network/MultiHead.hs#L26)
+
+## Decision
+
+* We implement our own message tracking and resending logic as a standalone `Network` layer
+* That layer consumes and produces `Authenticate`d messages in order to identify the source of messages
+* It uses a vector of monotonically increasing _sequence numbers_ associated with each party (including itself) to track what are the last messages from each party and to ensure FIFO delivery of messages
+* This _vector_ is also used to identify peers which are lagging behind and to resend the missing messages, or to drop messages which have already been received
+* This _vector_ is also piggybacked by the ping messages of the heartbeat mechanism so that a node informs its peers of its own situation even if it would have, otherwise, no message to sent to them
+* Sending a ping message does not influence the _vector_ of the sender
+* Any message received which index does not match what the peer expects is dropped
+* Messages deemed not received by a peer a re-sent
```

Comment:
```suggestion
* Messages deemed not received by a peer are re-sent
```

I would have expected this to be further up. For example as part of the first bullet point in the decision block